File: implementations/java-springboot/examples/src/main/java/com/example/accountservice/account/application/event/AccountSuspendedEventHandler.java
package com.example.accountservice.account.application.event;

import com.example.accountservice.account.application.integrationevent.AccountSuspendedIntegrationEventV1;
import com.example.accountservice.account.application.service.NotificationService;
import com.example.accountservice.account.domain.AccountSuspendedEvent;
import com.example.accountservice.outbox.OutboxEventHandler;
import com.example.accountservice.outbox.OutboxWriter;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

/**
 * Handles the {@link AccountSuspendedEvent} (an internal Domain Event) accumulated in the Outbox.
 *
 * <p>An EventHandler under application/event/ is the one exception allowed to use {@link
 * OutboxWriter} directly — here it converts the event into the Integration Event ({@code
 * account.suspended.v1}) exposed to external BCs (Card, etc.) and writes it into the same Outbox
 * transaction (an Aggregate never creates an Integration Event directly — the conversion point is
 * always the EventHandler). The newly written {@code account.suspended.v1} row is published to SQS
 * on {@code OutboxPoller}'s next polling tick and received/processed by {@code OutboxConsumer} — it
 * is not processed immediately within this handler's own invocation (it is asynchronous).
 */
@Component
@RequiredArgsConstructor
public class AccountSuspendedEventHandler implements OutboxEventHandler {

    private static final Logger log = LoggerFactory.getLogger(AccountSuspendedEventHandler.class);

    private final NotificationService notificationService;
    private final OutboxWriter outboxWriter;
    private final ObjectMapper objectMapper;

    @Override
    public String eventType() {
        return AccountSuspendedEvent.class.getSimpleName();
    }

    @Override
    public void handle(String payload) throws Exception {
        AccountSuspendedEvent event = objectMapper.readValue(payload, AccountSuspendedEvent.class);

        // Write the Integration Event notifying external BCs (Card, etc.) into the Outbox.
        outboxWriter.save(
                AccountSuspendedIntegrationEventV1.EVENT_TYPE,
                new AccountSuspendedIntegrationEventV1(event.accountId(), event.suspendedAt()));

        // The notification is best-effort. Even on failure, this method does not end by throwing —
        // throwing would cause this outbox row to be re-drained, which would write the Integration
        // Event above again (harmless since the receiver is idempotent, but this avoids unnecessary
        // amplification). Retrying the notification itself is handled by a separate outbox row.
        try {
            notificationService.sendEmail(
                    event.accountId(),
                    "AccountSuspended",
                    event.email(),
                    "[Account] Your account has been suspended",
                    "Account (" + event.accountId() + ") has been suspended.");
        } catch (Exception e) {
            log.error("Failed to send suspension notification", e);
        }
    }
}

File: implementations/kotlin-springboot/examples/src/main/kotlin/com/example/accountservice/account/application/event/AccountSuspendedEventHandler.kt
package com.example.accountservice.account.application.event

import com.example.accountservice.account.application.integrationevent.AccountSuspendedIntegrationEventV1
import com.example.accountservice.account.domain.AccountSuspendedEvent
import com.example.accountservice.notification.application.service.NotificationService
import com.example.accountservice.outbox.OutboxWriter
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

/**
 * Sends an account-suspension notification email when AccountSuspendedEvent is delivered via the Outbox,
 * and writes an Integration Event (account.suspended.v1) notifying external BCs (Card, etc.) into the
 * Outbox.
 *
 * The EventHandler in application/event/ is the one exception allowed to use [OutboxWriter] directly, and
 * this is where the internal Domain Event is converted into an Integration Event for external BCs (an
 * Aggregate never builds an Integration Event directly — the conversion point is always the
 * EventHandler). This handler itself is the callback that
 * [com.example.accountservice.outbox.OutboxConsumer] invokes when it receives AccountSuspendedEvent from
 * SQS — the Integration Event written here is not processed immediately within the same call; it is
 * published to SQS at the next [com.example.accountservice.outbox.OutboxPoller] tick (at most 1 second
 * later) and consumed separately (with the async transition, "immediate re-drain within the same
 * transaction" no longer holds).
 */
@Component
class AccountSuspendedEventHandler(
    private val notificationService: NotificationService,
    private val outboxWriter: OutboxWriter,
) {
    private val logger = LoggerFactory.getLogger(AccountSuspendedEventHandler::class.java)

    fun handle(
        event: AccountSuspendedEvent,
        eventId: String,
    ) {
        // Writes an Integration Event notifying external BCs (Card, etc.) into the Outbox.
        outboxWriter.saveAll(
            listOf(AccountSuspendedIntegrationEventV1(event.accountId, event.suspendedAt.toString())),
        )

        // The notification is best-effort. This handler is not allowed to throw on failure — if it did,
        // this Outbox row (AccountSuspendedEvent) itself would remain marked as failed and get
        // re-drained on the next call, and in that process the Integration Event already written above
        // would be written again (harmless since the receiving side is idempotent, but an unnecessary
        // amplification). Retrying the notification itself is the responsibility of a separate outbox
        // row (the sent_email pipeline). eventId is the eventId of this Outbox row — NotificationService
        // uses it as the key to apply Level 2 (Ledger) duplicate-send prevention.
        try {
            notificationService.sendEmail(
                accountId = event.accountId,
                eventType = "AccountSuspended",
                sourceEventId = eventId,
                recipient = event.email,
                subject = "[Account] Your account has been suspended",
                body = "Account (${event.accountId}) has been suspended.",
            )
        } catch (e: Exception) {
            logger
                .atError()
                .addKeyValue("account_id", event.accountId)
                .setCause(e)
                .log("Failed to send suspension notification")
        }
    }
}
